mcda/electre/concordance.py

```diff
@@ -1,617 +1,660 @@
+                raise TypeError("Interaction factor must be a numerical value.")
+            if (
+                interactions[i][j].interaction is not None
+                and interactions[i][j].interaction == "MW"
+                and weights[i] - abs(interactions[i][j].factor) < 0
+            ):
+                raise ValueError("Incorrect interaction factor.")
+            if (
+                interactions[i][j].interaction
+                and interactions[i][j].interaction == "A"
+                and weights[i] - interactions[i][j].factor < 0
+            ):
+                raise ValueError("Incorrect interaction factor.")
+
+    mutual_strengthening = []
+    mutual_weakening = []
+    antagonistic = []
+    for i in interactions.index:
+        for j in interactions.columns:
+            if interactions[i][j].interaction is not None:
+                c_i = concordance_marginal(
+                    a_values[i],
+                    b_values[i],
+                    scales[i],
+                    indifference_thresholds[i],
+                    preference_thresholds[i],
+                )
+                c_j = concordance_marginal(
+                    a_values[j],
+                    b_values[j],
+                    scales[j],
+                    indifference_thresholds[j],
+                    preference_thresholds[j],
+                )
+                if interactions[i][j].interaction == "MS":
+                    strengthening_weight = (
+                        weights[i] + weights[j] + interactions[i][j].factor
+                    )
+                    mutual_strengthening.append(
+                        strengthening_weight * min(c_i, c_j)
+                        if interactions[i][j].function == "min"
+                        else strengthening_weight * c_i * c_j
+                    )
+                elif interactions[i][j].interaction == "MW":
+                    weakening_weight = weights[i] + weights[j] + interactions[i][j].factor
+                    mutual_weakening.append(
+                        weakening_weight * min(c_i, c_j)
+                        if interactions[i][j].function == "min"
+                        else weakening_weight * c_i * c_j
+                    )
+                else:
+                    antagonistic_weight = (
+                        weights[i] + weights[j] - interactions[i][j].factor
+                    )
+                    antagonistic.append(
+                        antagonistic_weight * min(c_i, c_j)
+                        if interactions[i][j].function == "min"
+                        else antagonistic_weight * c_i * c_j
+                    )
+
+    return (
+        sum(
+            [
+                weights[i]
+                * concordance_marginal(
+                    a_values[i],
+                    b_values[i],
+                    scales[i],
+                    indifference_thresholds[i],
+                    preference_thresholds[i],

```

Comment:
there are multiple calculations of the same marginal concordance indices - please optimize this